src/backend/access/transam/csn_snapshot.c

```diff
@@ -0,0 +1,408 @@
+ * GenerateCSN().
+ */
+typedef struct
+{
+	SnapshotCSN		 last_max_csn;		/* Record the max csn till now */
+	XidCSN			 last_csn_log_wal;	/* for interval we log the assign csn to wal */
+	TransactionId 	 xmin_for_csn; 		/*'xmin_for_csn' for when turn xid-snapshot to csn-snapshot*/
+	volatile slock_t lock;
+} CSNSnapshotState;
+
+static CSNSnapshotState *csnState;
+
+/*
+ * Enables this module.
+ */
+extern bool enable_csn_snapshot;
+
+
+/* Estimate shared memory space needed */
+Size
+CSNSnapshotShmemSize(void)
+{
+	Size	size = 0;
+
+	size += MAXALIGN(sizeof(CSNSnapshotState));
+
+	return size;
+}
+
+/* Init shared memory structures */
+void
+CSNSnapshotShmemInit()
+{
+	bool found;
+
+	csnState = ShmemInitStruct("csnState",
+							sizeof(CSNSnapshotState),
+							&found);
+	if (!found)
+	{
+		csnState->last_max_csn = 0;
+		csnState->last_csn_log_wal = 0;
+		SpinLockInit(&csnState->lock);
+	}
+}
+
+/*
+ * GenerateCSN
+ *
+ * Generate SnapshotCSN which is actually a local time. Also we are forcing
+ * this time to be always increasing. Since now it is not uncommon to have
+ * millions of read transactions per second we are trying to use nanoseconds
+ * if such time resolution is available.
+ */
+SnapshotCSN
+GenerateCSN(bool locked)
+{
+	instr_time	current_time;
+	SnapshotCSN	csn;
+
+	Assert(get_csnlog_status());
+
+	/*
+	 * TODO: create some macro that add small random shift to current time.
+	 */
+	INSTR_TIME_SET_CURRENT(current_time);
+	csn = (SnapshotCSN) INSTR_TIME_GET_NANOSEC(current_time);
+
+	/* TODO: change to atomics? */
+	if (!locked)
+		SpinLockAcquire(&csnState->lock);
+
+	if (csn <= csnState->last_max_csn)
+		csn = ++csnState->last_max_csn;
+	else
+		csnState->last_max_csn = csn;
+
+	WriteAssignCSNXlogRec(csn);
+
+	if (!locked)
+		SpinLockRelease(&csnState->lock);
+
+	return csn;
+}
+
+/*
+ * TransactionIdGetXidCSN
+ *
+ * Get XidCSN for specified TransactionId taking care about special xids,
+ * xids beyond TransactionXmin and InDoubt states.
+ */
+XidCSN
+TransactionIdGetXidCSN(TransactionId xid)
+{
+	XidCSN 			 xid_csn;
+
+	Assert(get_csnlog_status());
+
+	/* Handle permanent TransactionId's for which we don't have mapping */
+	if (!TransactionIdIsNormal(xid))
+	{
+		if (xid == InvalidTransactionId)
+			return AbortedXidCSN;
+		if (xid == FrozenTransactionId || xid == BootstrapTransactionId)
+			return FrozenXidCSN;
+		Assert(false); /* Should not happend */
+	}
+
+	/*
+	 * If we just switch a xid-snapsot to a csn_snapshot, we should handle a start
+	 * xid for csn basse check. Just in case we have prepared transaction which
+	 * hold the TransactionXmin but without CSN.
+	 */
+	if(InvalidTransactionId == xmin_for_csn)
+	{
+		SpinLockAcquire(&csnState->lock);
+		if(InvalidTransactionId != csnState->xmin_for_csn)
+			xmin_for_csn = csnState->xmin_for_csn;
+		else
+			xmin_for_csn = FrozenTransactionId;
+
+		SpinLockRelease(&csnState->lock);
+	}
+
+	if ( FrozenTransactionId != xmin_for_csn ||
```

Comment:
This should be `FrozenTransactionId == xmin_for_csn`

Reply:
This was a part of the change in the second part of the patch -- Patch 0002-Global-snapshots 

I think this works because the second part of the if condition will be true. i.e, TransactionIdPrecedes(xmin_for_csn, TransactionXmin) because xmin_for_csn < FirstNormalTransactionId 

Let me know if you think I am missing something.